### ISSUE TITLE
Crash when loaded together with Polymake.jl

### ISSUE THREAD

**Issue by saschatimme** [Nemocas/Nemo.jl#788]
Using the current Polymake.jl master (sha 2ea6decd9c607bc52088a7ac83824269117b4ff7) with the latest release of Nemo.jl results in this crash:

```
julia> using Polymake;
polymake version 4.0
Copyright (c) 1997-2020
Ewgenij Gawrilow, Michael Joswig, and the polymake team
Technische Universität Berlin, Germany
https://polymake.org

This is free software licensed under GPL; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

julia> C = polytope.cube(3);
julia> using Nemo;

Welcome to Nemo version 0.16.2

Nemo comes with absolutely no warranty whatsoever


julia> CC, s = Nemo.PolynomialRing(ComplexField(256), "s");

julia> s
[ 0julia(50132,0x109193dc0) malloc: *** error for object 0x7f8501ce99a0: pointer being freed was not allocated
julia(50132,0x109193dc0) malloc: *** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
in expression starting at REPL[5]:0
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 29021319 (Pool: 29013109; Big: 8210); GC: 29
zsh: abort      julia
```
The crash does not happen if Nemo and Polymake are loaded in the other order.

Before we upgraded to polymake 4.0 and a new binary builder version things worked. To the previous version it changed that we now also include the binary builder `flint` version (https://github.com/thofma/Flint2Builder/releases/tag/ba0cee).

We are not sure, whether we messed up on the polymake side or whether it is an issue with the code running in `__init__` in Nemo. 
Can you help us out here? 

cc: @benlorenz, @kalmarek 

P.S.: I accidentally opened this issue first in the wrong repo https://github.com/wbhart/Nemo.jl/issues/108

**Comment by wbhart** [comment]
We can't do much here without a proper Julia stacktrace. Can you prevent Polymake taking over that?

**Comment by thofma** [comment]
Can you post the `deps/deps.jl` somewhere? It is created by the build script and not included in the Polymake.jl repository.

Also, are you setting the *alloc functions of flint to use the julia ones? If not, you should try this.

**Comment by kalmarek** [comment]
below is my `deps/deps.jl`.
```julia
## This file autogenerated by BinaryProvider.write_deps_file().
## Do not edit.
##
## Include this file within your main top-level source, and call
## `check_deps()` from within your module's `__init__()` method

if isdefined((@static VERSION < v"0.7.0-DEV.484" ? current_module() : @__MODULE__), :Compat)
    import Compat.Libdl
elseif VERSION >= v"0.7.0-DEV.3382"
    import Libdl
end
const libpolymake = joinpath(dirname(@__FILE__), "usr/lib/libpolymake-apps-rt.so")
const polymake = joinpath(dirname(@__FILE__), "usr/bin/polymake")
const polymake_config = joinpath(dirname(@__FILE__), "usr/bin/polymake-config")
const ninja = joinpath(dirname(@__FILE__), "usr/bin/ninja")
const perl = joinpath(dirname(@__FILE__), "usr/bin/perl")
function check_deps()
    global libpolymake
    if !isfile(libpolymake)
        error("$(libpolymake) does not exist, Please re-run Pkg.build(\"Polymake\"), and restart Julia.")
    end

    if Libdl.dlopen_e(libpolymake) in (C_NULL, nothing)
        error("$(libpolymake) cannot be opened, Please re-run Pkg.build(\"Polymake\"), and restart Julia.")
    end

    global polymake
    if !isfile(polymake)
        error("$(polymake) does not exist, Please re-run Pkg.build(\"Polymake\"), and restart Julia.")
    end

    global polymake_config
    if !isfile(polymake_config)
        error("$(polymake_config) does not exist, Please re-run Pkg.build(\"Polymake\"), and restart Julia.")
    end

    global ninja
    if !isfile(ninja)
        error("$(ninja) does not exist, Please re-run Pkg.build(\"Polymake\"), and restart Julia.")
    end

    global perl
    if !isfile(perl)
        error("$(perl) does not exist, Please re-run Pkg.build(\"Polymake\"), and restart Julia.")
    end

    libpaths = split(get(ENV, "LD_LIBRARY_PATH", ""), ":")
    if !("/opt/julia-1.3.0/bin/../lib/julia" in libpaths)
        push!(libpaths, "/opt/julia-1.3.0/bin/../lib/julia")
    end
    ENV["LD_LIBRARY_PATH"] = join(filter(!isempty, libpaths), ":")

end
const using_binary = true
begin
    #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:99 =#
    using Pkg: depots1
    #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:100 =#
    function prepare_env()
        #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:101 =#
        ENV["PERL5LIB"] = "/home/kalmar/.julia/dev/Polymake/deps/usr/lib/perl5/site_perl/5.30.0/x86_64-linux:/home/kalmar/.julia/dev/Polymake/deps/usr/lib/perl5/site_perl/5.30.0:/home/kalmar/.julia/dev/Polymake/deps/usr/lib/perl5/5.30.0/x86_64-linux:/home/kalmar/.julia/dev/Polymake/deps/usr/lib/perl5/5.30.0"
        #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:102 =#
        user_dir = (ENV["POLYMAKE_USER_DIR"] = abspath(joinpath(depots1(), "polymake_user")))
        #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:103 =#
        if Base.Filesystem.isdir(user_dir)
            #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:104 =#
            del = filter((i->begin
                            #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:104 =#
                            Base.Filesystem.isdir(i) && startswith(i, "wrappers.")
                        end), readdir(user_dir))
            #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:105 =#
            for i = del
                #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:106 =#
                Base.Filesystem.rm(user_dir * "/" * i, recursive=true)
            end
        end
        #= /home/kalmar/.julia/dev/Polymake/deps/build.jl:109 =#
        ENV["PATH"] = ENV["PATH"] * ":" * "/home/kalmar/.julia/dev/Polymake/deps/usr" * "/bin"
    end
end
```

**Comment by thofma** [comment]
Thanks. Did you try setting the malloc and friends functions as suggested?

**Comment by benlorenz** [comment]
Setting the malloc function requires some extra detours as we are not dlopen'ing libflint directly but it is loaded automatically as a library-dependency for libpolymake:
Loading libflint.so manually (with a full path) and calling `:__flint_set_memory_functions` before loading Polymake.jl fixes the crash.

PS: stacktrace from linux, without setting the memory functions:
```
signal (6): Aborted
in expression starting at REPL[6]:0
gsignal at /lib64/libc.so.6 (unknown line)
abort at /lib64/libc.so.6 (unknown line)
__libc_message at /lib64/libc.so.6 (unknown line)
malloc_printerr at /lib64/libc.so.6 (unknown line)
munmap_chunk at /lib64/libc.so.6 (unknown line)
jl_gc_counted_free_with_size at /tmp/julia-1.3.1/work/julia-1.3.1/src/gc.c:3040
show at $HOME/.julia/packages/Nemo/LBWHq/src/arb/arb.jl:147
show at $HOME/.julia/packages/Nemo/LBWHq/src/arb/acb.jl:175
print at ./strings/io.jl:35
show at $HOME/.julia/packages/Nemo/LBWHq/src/arb/acb_poly.jl:90
show at ./multimedia.jl:47
display at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:132
display at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:136
display at ./multimedia.jl:323
jl_apply at /tmp/julia-1.3.1/work/julia-1.3.1/src/julia.h:1631 [inlined]
jl_f__apply at /tmp/julia-1.3.1/work/julia-1.3.1/src/builtins.c:627
jl_f__apply_latest at /tmp/julia-1.3.1/work/julia-1.3.1/src/builtins.c:665
#invokelatest#1 at ./essentials.jl:709 [inlined]
invokelatest at ./essentials.jl:708 [inlined]
print_response at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:156
unknown function (ip: 0x7f3c36870d89)
print_response at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:141
unknown function (ip: 0x7f3c36870485)
do_respond at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:719
unknown function (ip: 0x7f3c36896d33)
jl_apply at /tmp/julia-1.3.1/work/julia-1.3.1/src/julia.h:1631 [inlined]
jl_f__apply at /tmp/julia-1.3.1/work/julia-1.3.1/src/builtins.c:627
jl_f__apply_latest at /tmp/julia-1.3.1/work/julia-1.3.1/src/builtins.c:665
#invokelatest#1 at ./essentials.jl:709 [inlined]
invokelatest at ./essentials.jl:708 [inlined]
run_interface at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/LineEdit.jl:2306
run_frontend at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:1045
run_repl at /tmp/julia-1.3.1/work/julia-1.3.1/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:201
#770 at ./client.jl:382
jfptr_#770_7646 at /usr/lib64/julia/sys.so (unknown line)
jl_apply at /tmp/julia-1.3.1/work/julia-1.3.1/src/julia.h:1631 [inlined]
jl_f__apply at /tmp/julia-1.3.1/work/julia-1.3.1/src/builtins.c:627
jl_f__apply_latest at /tmp/julia-1.3.1/work/julia-1.3.1/src/builtins.c:665
#invokelatest#1 at ./essentials.jl:709 [inlined]
invokelatest at ./essentials.jl:708 [inlined]
run_main_repl at ./client.jl:366
exec_options at ./client.jl:304
_start at ./client.jl:460
jfptr__start_2084 at /usr/lib64/julia/sys.so (unknown line)
unknown function (ip: 0x55b71040b755)
unknown function (ip: 0x55b71040b332)
__libc_start_main at /lib64/libc.so.6 (unknown line)
unknown function (ip: 0x55b71040b3d9)
Allocations: 23273266 (Pool: 23268330; Big: 4936); GC: 18
Aborted
```

**Comment by fieker** [comment]
this is random - I encountered this yesterday extensively. It's "fixed" in 1.4.0-rc2
It should be fixable with Nemo/Hecke changes (specifying the libraries explicitly), however it did not work.
My guess is that the resolving of gmp inside flint is "random". I can specify the libflint and that will be honored, but spcifying the gmp used by flint is hard

**Comment by wbhart** [comment]
It's supposed to do that via RPATH, I thought.

**Comment by micjoswig** [comment]
Let me add a general comment concerning the structure of OSCAR.

The collaboration among the cornerstone systems has a long (and successful) history by now.  Hence there are many mutual includes and references.  This was OK as long as the systems existed mostly independently.  Now, on the OSCAR level, in Julia, this is no longer feasible.

Some of you may recall our week of coding and discussions in Berlin in the fall of 2018.  I advocated a three layer model, where the first layer are the libraries, the second are the cornerstones, and the third one is the actual OSCAR.

In this setup flint counts as a library, and as such it should be loaded first, with whatever parameters the gods of flint deem useful.  The corner stones (on the second layer) call, e.g., flint with whatever parameters were chosen before, and they don't tweak them.  Actually, there is no other way, as it is impossible in Julia to load the same library twice (and for good reason). Tweaking paths and symbols is not a good idea.  It only looks like a way out, but it is shortsighted and impossible to maintain in the long run.

There are other examples: Singular calls Nemo, GAP calls polymake, and polymake calls Singular.  None of this will remain feasible on the OSCAR level.  Yes, this will require the developers of all four cornerstones to change their systems.  No cornerstone should call any other cornerstone directly.  Otherwise we will have cyclic dependences. The communication needs to happen on the third layer, OSCAR. 

We do want to keep the cornerstones as independent systems.  As such they can call whatever they want.  But the dependence on the other cornerstones must be switched off in the Julia interfaces.  There is no other way.

One of the big advantages of Julia is the package manager and the build system.  We should make use of it, not fight against it.

PS: Wouldn't it be nice if Nemo would be added to https://github.com/oscar-system?

**Comment by fieker** [comment]
On Thu, Feb 27, 2020 at 04:36:07AM -0800, wbhart wrote:
> It's supposed to do that via RPATH, I thought.
> 
it's also a question of when
I load libflint - but do not resolve anything(?)
Then I call a flint function and trigger the correct gmp
thus a flint function is now calling the correct gmp

Then polymake is added and thus a 2nd flint and a 2nd gmp are added

Now I call the 1st flint again which calls a new gmp function. Will this
now be resolved from the 1st gmp or the 2nd?

Looking at the dlopen and the libdl.so doc, it is not clear what the
linker is supposed to be doing here. At this point, the rpath is no
longer an issue

**Comment by micjoswig** [comment]
> Then polymake is added and thus a 2nd flint and a 2nd gmp are added Now I call the 1st flint again which calls a new gmp function. Will this now be resolved from the 1st gmp or the 2nd? Looking at the dlopen and the libdl.so doc, it is not clear what the linker is supposed to be doing here. At this point, the rpath is no longer an issue

As far as I understand there is no such thing as loading twice.  The second load just does not happen.

**Comment by wbhart** [comment]
There's a binary builder for Flint which you can use, here, without adding a dependency on Nemo:

https://github.com/thofma/Flint2Builder/releases

Then you configure Polymake (in the BinaryBuilder for Polymake.jl) to build against that Flint.

You have to do a similar thing with GMP, for which Julia provides a BinaryBuilder:

https://github.com/JuliaMath/GMPBuilder/releases

However, you still have to set the GMP and Flint memory manager functions as we do in Nemo, in Polymake.jl.

**Comment by wbhart** [comment]
There's a binary builder for Flint which you can use, here, without adding a dependency on Nemo:

https://github.com/thofma/Flint2Builder/releases

Then you configure Polymake (in the BinaryBuilder for Polymake.jl) to build against that Flint.

You have to do a similar thing with GMP, for which Julia provides a BinaryBuilder:

https://github.com/JuliaMath/GMPBuilder/releases

However, you still have to set the GMP and Flint memory manager functions as we do in Nemo, in Polymake.jl.

**Comment by kalmarek** [comment]
https://github.com/benlorenz/polymakeBuilder/blob/3443aac0521a6b2e71fdc3c673f1b542f2eb9d2d/build_tarballs.jl#L121

I don't think that Polymake.jl should be doing any fine tuning for its external dependencies?
But as a workaround could you suggest a way of keeping them in sync between Polymake.jl and Nemo.jl without hand copy-pasting?

**Comment by fieker** [comment]
On Thu, Feb 27, 2020 at 05:43:18AM -0800, Michael Joswig wrote:
> Let me add a general comment concerning the structure of OSCAR.
> 
> The collaboration among the cornerstone systems has a long (and
> successful) history by now.  Hence there are many mutual includes and
> references.  This was OK as long as the systems existed mostly
> independently.  Now, on the OSCAR level, in Julia, this is no longer
> feasible.
> 
> Some of you may recall our week of coding and discussions in Berlin in
> the fall of 2018.  I advocated a three layer model, where the first
> layer are the libraries, the second are the cornerstones, and the
> third one is the actual OSCAR.
> 
> In this setup flint counts as a library, and as such it should be
> loaded first, with whatever parameters the gods of flint deem useful.
Is happening....
> The corner stones (on the second layer) call, e.g., flint with
> whatever parameters were chosen before, and they don't tweak them.
Is not happening...
> Actually, there is no other way, as it is impossible in Julia to load
> the same library twice (and for good reason). Tweaking paths and
Oh, yes, the same library is loaded twice. I attached the debugger and
set breakpoints. Clearly, in Oscar we have both libgmp and libflint
available twice.
> symbols is not a good idea.  It only looks like a way out, but it is
> shortsighted and impossible to maintain in the long run.
I agree.

The problem (one of them) is the allocators.
I presume polymake uses flint (and gmp), both libraries allocate memory
for results.
Problem is the the identical library needs to do the free.
And here we're in trouble
Sequence of calls:
  alloc:
    antic(nemo) -> flint(polymake) -> gmp(polymake) -> alloc(system)

    nf_elem_get_str_pretty -> _fmpq_poly_get_str_pretty -> flint_malloc
                                                               -> malloc
  free
    flint_free(nemo) -> free(julia)

I can only control the top layer (as far as I can see), ie. I can make
sure I call antic(nemo) and flint(nemo). What flint/ antic are then calling
depends on the libdl - which seems to be a bit random, possibly on
purpose.

Technically, it should be possible to ask the dlopen for flint to
immediately resolve all calls to gmp to the (at this time) only (and
correct) copy, but my experiments were not successful

Alternatively, Polymake.jl could set the allocater/free functions in
their copy of flint/gmp to the same functions as Nemo...

Alternatively, we could try to figure out how to use only one copy of
flint/gmp with only one set of memory functions...
> 
> There are other examples: Singular calls Nemo, GAP calls polymake, and
> polymake calls Singular.  None of this will remain feasible on the
> OSCAR level.  Yes, this will require the developers of all four
> cornerstones to change their systems.  No cornerstone should call any
> other cornerstone directly.  Otherwise we will have cyclic
> dependences. The communication needs to happen on the third layer,
> OSCAR. 
So, you are going to remove flint from polymake? What about gmp?
> 
> We do want to keep the cornerstones as independent systems.  As such
> they can call whatever they want.  But the dependence on the other
> cornerstones must be switched off in the Julia interfaces.  There is
> no other way.
The problem here is not the dependence of the cornerstones, the problem
are the overlap on the c-level 
For Singular we made it use the Nemo gmp and flint - which is part of
the reason to have the dependency
For Gap we're trying the "same": make Gap use the Nemo gmp (or the julia
one: both Nemo and julia use the same memory manager)
> 
> One of the big advantages of Julia is the package manager and the
> build system.  We should make use of it, not fight against it.
Yep, definitely.

So, provocatively: who is allowed to directly load (and use) flint?
Nemo, Singular or Polymake?
Currently they all three do...

**Comment by fieker** [comment]
On Thu, Feb 27, 2020 at 05:49:12AM -0800, Michael Joswig wrote:
> > Then polymake is added and thus a 2nd flint and a 2nd gmp are added Now I call the 1st flint again which calls a new gmp function. Will this now be resolved from the 1st gmp or the 2nd? Looking at the dlopen and the libdl.so doc, it is not clear what the linker is supposed to be doing here. At this point, the rpath is no longer an issue
> 
> As far as I understand there is no such thing as loading twice.  The second load just does not happen.

It really does happen, gdb showed me two copys, lsof showed two
copies...

**Comment by wbhart** [comment]
@kalmarek Which fine tuning are you talking about? Flint has no tuning options and GMP tuning is supplied with the library.

**Comment by wbhart** [comment]
Polymake.jl has to set the Flint memory manager to the Julia one at runtime, otherwise Polymake will crash when the gc runs out of memory. (Probably does anyway, but that is a Julia gc bug that will eventually be fixed by the Julia people, I trust.)

**Comment by micjoswig** [comment]
> So, you are going to remove flint from polymake? What about gmp?

No, of course, not flint and gmp are both libraries.

> We do want to keep the cornerstones as independent systems. As such they can call whatever they want. But the dependence on the other cornerstones must be switched off in the Julia interfaces. There is no other way.
> The problem here is not the dependence of the cornerstones, the problem are the overlap on the c-level For Singular we made it use the Nemo gmp and flint - which is part of the reason to have the dependency For Gap we're trying the "same": make Gap use the Nemo gmp (or the julia one: both Nemo and julia use the same memory manager)

Singular calling Nemo is the problem here, unless Nemo is a library. In my understanding Nemo is a cornerstone. No?

**Comment by kalmarek** [comment]
> So, provocatively: who is allowed to directly load (and use) flint?
Nemo, Singular or Polymake?
Currently they all three do...

My provocative guess: none of them ;), since we're aiming at `fmpz` based Integers in the whole system, libflint should be ready with all of its setup before anything else.

> > As far as I understand there is no such thing as loading twice. The second load just does not happen.
> It really does happen, gdb showed me two copys, lsof showed two copies...

Oh, my mistake then, but given what @benlorenz wrote 2 weeks ago about manual loading, the first copy gets precedence (i.e. it is searched for symbols and ccalled)?

> Which fine tuning are you talking about? Flint has no tuning options and GMP tuning is supplied with the library.

I didn't mean 'performance tuning', what I meant was that we use julia GMP 'as is', without setting any functions (despite the fact that julia overwrites eg. its `free`). We had some crashes due to bad frees, but these has been since fixed in Julia. But I'm no expert, @benlorenz is definitely more competent on this. 

> My understanding is that you are interacting with Flint at a lower level than what Singular is. It does it via the high level Nemo interface. You do it through a low level C/C++ interface. Correct me if I'm wrong about that.

That is very much correct.

> Nemo could just as well be called Flint.jl, as its main purpose is to provide a Julia interface to Flint.

As I understand Nemo serves two functions: providing interface to Flint AND defining algebraic structures upon data structures from Flint. This makes it different from Polymake.jl, which does only the first thing (that's why Nemo.jl needs AbstractAlgebra.jl and Polymake.jl doesn't).

**Comment by fieker** [comment]
On Thu, Feb 27, 2020 at 02:37:53PM +0000, Michael Joswig wrote:
> > So, you are going to remove flint from polymake? What about gmp?
> 
> No, of course, not flint and gmp are both libraries.
> 
> > We do want to keep the cornerstones as independent systems. As such they can call whatever they want. But the dependence on the other cornerstones must be switched off in the Julia interfaces. There is no other way.
> > The problem here is not the dependence of the cornerstones, the problem are the overlap on the c-level For Singular we made it use the Nemo gmp and flint - which is part of the reason to have the dependency For Gap we're trying the "same": make Gap use the Nemo gmp (or the julia one: both Nemo and julia use the same memory manager)
> 
> Singular calling Nemo is the problem here, unless Nemo is a library. In my understanding Nemo is a cornerstone. No?
Singular is not part of the problem. Its just between Nemo and Polymake
as bot rely on gmp and flint

**Comment by fieker** [comment]
On Thu, Feb 27, 2020 at 07:27:38AM -0800, kalmarek wrote:
> > So, provocatively: who is allowed to directly load (and use) flint?
> Nemo, Singular or Polymake?
> Currently they all three do...
> 
> My provocative guess: none of them ;), since we're aiming at `fmpz` based Integers in the whole system, libflint should be ready with all of its setup before anything else.
> 
> > > As far as I understand there is no such thing as loading twice. The second load just does not happen.
> > It really does happen, gdb showed me two copys, lsof showed two copies...
> 
> Oh, my mistake then, but given what @benlorenz wrote 2 weeks ago about manual loading, the first copy gets precedence (i.e. it is searched for symbols and ccalled)?
> 
I tried to understand the libdl and there is a depth first search,
definetely not 1st is searched first, that would be to easy

**Comment by micjoswig** [comment]
OK, then Nemo actually works as a library, and then Polymake.jl should go through Nemo to access gmp and and flint.  It is probably my fault that I did not get this right.  No problem, we will fix this then.

**Comment by wbhart** [comment]
@micjoswig At least that solution should work for you, so long as everything is still in the same process.

By the way, Nemo doesn't provide access to GMP per se, Julia does (though of course Flint internally uses GMP).

The reasons why the memory managers have to be in sync is technical, and not well understood by any of us. Fundamentally it does come down to objects being allocated by one system and freed by another.

There's probably more than one reason, e.g.. Flint objects being allocated by Nemo/Julia then passed to Flint, then swapped internally for Flint allocated objects, or objects are allocated by Flint then passed back to Julia with the expectation that the user clean them up.

Either way, we've developed a mechanism that works. It just requires everyone to use the same memory manager.

Currently if you want to use threading, you have to turn it all off anyway, which is why it is a runtime option. That's due to a bug in Julia.

**Comment by fieker** [comment]
On Tue, Mar 10, 2020 at 05:34:28AM -0700, Sascha Timme wrote:
> Closed #788 via oscar-system/Polymake.jl#232.

Great - thanks - weill see the effect on Travis...
